File: dcp-solutions/2024-12-14/solution.php
```php
class Solution {
    public function nextPermutation($nums) {
        $n = count($nums);
        $i = $n - 2;
        
        while ($i >= 0 && $nums[$i] >= $nums[$i + 1]) {
            $i -= 1;
        }
        
        if ($i < 0) return $nums;
        
        $j = $n - 1;
        while ($nums[$j] <= $nums[$i]) {
            $j -= 1;
        }
        
        $tmp = $nums[$i];
        $nums[$i] = $nums[$j];
        $nums[$j] = $tmp;
        
        $left = $i + 1;
        $right = $n - 1;
        
        while ($left < $right) {
            $tmp = $nums[$left];
            $nums[$left] = $nums[$right];
            $nums[$right] = $tmp;
            $left += 1;
            $right -= 1;
        }
    }
}
```

File: dcp-solutions/2024-12-14/solution.c
```c
#include <stdio.h>

void nextGreaterElement(int* nums, int numsSize, int* returnSize) {
    if (numsSize == 0) *returnSize = 0;
    
    int i = numsSize - 2;
    while (i >= 0 && nums[i] >= nums[i + 1]) i--;
    
    if (i < 0) {
        *returnSize = 1;
        for(int j=0; j<numsSize-1; j++) (*returnSize)*10+nums[j], nums[j]=0;
        nums[numsSize-1] = 1;
        return;
    }
    
    int j = numsSize - 1;
    while (nums[i] >= nums[j]) j--;
    
    swap(nums[i], nums[j]);
    int left = i + 1, right = numsSize - 1;
    while(left < right) {
        swap(nums[left], nums[right]);
        left++, right--;
    }
}

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}
```